third_party/move/move-vm/runtime/src/loader/mod.rs

```diff
@@ -793,12 +824,134 @@ impl Loader {
+            let (module, size) = match module_store.module_at_by_ref(&ModuleIdRef::new(addr, name))
+            {
+                Some(module) => (module.module.clone(), module.size),
+                None => {
+                    let (module, size, _) = data_store.load_compiled_module_to_cache(
+                        ModuleId::new(*addr, name.to_owned()),
+                        allow_loading_failure,
+                    )?;
+                    (module, size)
+                },
+            };
+
+            // Extend the lifetime of the module to the remainder of the function body
+            // by storing it in an arena.
+            //
+            // This is needed because we need to store references derived from it in the
+            // work list.
+            let module = referenced_modules.alloc(module);
+
+            gas_meter
+                .charge_dependency(false, addr, name, NumBytes::new(size as u64))
+                .map_err(|err| {
+                    err.finish(Location::Module(ModuleId::new(*addr, name.to_owned())))
+                })?;
+
+            // Explore all dependencies and friends that have been visited yet.
+            for (addr, name) in module
+                .immediate_dependencies_iter()
+                .chain(module.immediate_friends_iter())
+                .rev()
+            {
+                if visited.insert((addr, name), ()).is_none() {
+                    stack.push((addr, name, false));
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Similar to `check_dependencies_and_charge_gas`, except that this does not recurse
+    /// into transitive dependencies and allows non-existent modules.
+    pub(crate) fn check_dependencies_and_charge_gas_non_recursive_optional<'a, I>(
+        &self,
+        module_store: &ModuleStorageAdapter,
+        data_store: &mut TransactionDataCache,
+        gas_meter: &mut impl GasMeter,
+        visited: &mut BTreeMap<(&'a AccountAddress, &'a IdentStr), ()>,
+        ids: I,
+    ) -> VMResult<()>
+    where
+        I: IntoIterator<Item = (&'a AccountAddress, &'a IdentStr)>,
+    {
+        for (addr, name) in ids.into_iter() {
+            if visited.insert((addr, name), ()).is_some() {
+                continue;
+            }
+
+            // Load and deserialize the module only if it has not been cached by the loader.
+            // Otherwise this will cause a significant regression in performance.
+            let size = match module_store.module_at_by_ref(&ModuleIdRef::new(addr, name)) {
+                Some(module) => module.size,
+                None => match data_store
+                    .load_compiled_module_to_cache(ModuleId::new(*addr, name.to_owned()), true)
+                {
+                    Ok((_module, size, _hash)) => size,
+                    Err(err) if err.major_status() == StatusCode::LINKER_ERROR => continue,
```

Comment:
Why? Linker error happens if a module does not exist, but this is not really possible, why not propagate error all the way?

Reply:
I forgot the details, but the loader is currently doing some funky error conversion, which gets broken if I remove this line.